examples/audio/lhotse_example.py

```diff
@@ -0,0 +1,235 @@
+
+
+class RecombineBatchAfterIO(dp.IterDataPipe):
+    def __init__(self, datapipe):
+        self.datapipe = datapipe
+
+    def __iter__(self):
+        # This is a buffer that will be re-combining batches.
+        # We might possibly get cuts from the same batch out-of-order.
+        batches: Dict[int, List[Cut]] = defaultdict(list)
+        for cut_idx, batch_idx, batch_size, audio, cut in self.datapipe:
+            batches[batch_idx].append((audio, cut))
+            if len(batches[batch_idx]) == batch_size:
+                audios, cuts = zip(*batches[batch_idx])
+                yield audios, CutSet.from_cuts(cuts)
+                # Free up the buffer dict after using the batch.
+                del batches[batch_idx]
+
+
+class UnbatchForIO(dp.IterDataPipe):
+    def __init__(self, datapipe):
+        self.datapipe = datapipe
+
+    def __iter__(self):
+        for batch_idx, cuts_batch in enumerate(self.datapipe):
+            batch_size = len(cuts_batch)
+            for cut_idx, cut in enumerate(cuts_batch):
+                yield cut_idx, batch_idx, batch_size, cut
+
+
+def classifier_fn(value, num_jobs):
+    return value[0] % num_jobs
+
+
+def build_pipeline(path: str, num_jobs: int = 2):
+
+    # Open cuts manifest for lazy reading
+    datapipe = CutsReader(path)
+
+    # Custom Sampling
+    datapipe = DurationBatcher(datapipe, max_duration=100)
+
+    # Unbatch the batched to run I/O
+    datapipe = UnbatchForIO(datapipe)  # Yield (batch index, dataclass)
+
+    # The following Multiprocessing API may change, and multiprocessing should be added and handled by DataLoader
+    # demux and mux would be called by DataLoader. But, I want to explicitly call them to illustrate the functionality
+
+    # Routed prior datapipe to ``num_jobs`` datapipes in round-robin manner
+    datapipes = datapipe.demux(
+        num_jobs, classifier_fn=partial(classifier_fn, num_jobs=num_jobs)
+    )
+    for i in range(len(datapipes)):
+        datapipes[i] = IODataPipe(datapipes[i])
+    import multiprocessing
+
+    ctx = multiprocessing.get_context("spawn")
+    for i in range(num_jobs):
+        process, req_queue, res_queue = SpawnProcessForDataPipeline(ctx, datapipes[i])
+        process.start()
+        datapipes[i] = QueueWrapper(
+            IterDataPipeQueueProtocolClient(req_queue, res_queue)
+        )
+    # DataLoader will also handle stop, join and cleanup subprocesses
+
+    datapipe = dp.Multiplexer(*datapipes)
+
+    datapipe = RecombineBatchAfterIO(datapipe)
+    # datapipe = datapipe.collate().tranform(transforms=...).filter(filter_fn=...)
+    return datapipe
+
+
+if __name__ == "__main__":
+    cuts_path = Path(
+        "./workspace/LibriSpeech/manifests/mini-libri-cuts-train-clean-5.jsonl.gz"
+    )
+
+    if not cuts_path.exists():
+        print("Downloading Mini LibriSpeech.")
+        download_librispeech("./workspace", dataset_parts="mini_librispeech")
```

Comment:
This could potentially replaced by the `on_disk_cache` [WIP for refactoring] with download and hash check.